packages/node/src/lib/token-manager.ts

```diff
@@ -0,0 +1,290 @@
+
+  constructor(props: OAuthSettings) {
+    this.keyId = props.keyId
+    this.clientId = props.clientId
+    this.clientKey = props.clientKey
+    this.authServer = props.authServer ?? 'https://oauth2.segment.io'
+    this.scope = props.scope ?? 'tracking_api:write'
+    this.httpClient = props.httpClient ?? new FetchHTTPClient()
+    this.maxRetries = props.maxRetries ?? 3
+    this.tokenEmitter.on('access_token', (event) => {
+      if ('token' in event) {
+        this.accessToken = event.token
+      }
+    })
+    this.controller = new AbortController()
+    this.signal = this.controller.signal
+    this.signal.addEventListener('abort', () => {
+      clearTimeout(this.pollerTimer)
+    })
+    this.retryCount = 0
+  }
+
+  async pollerLoop() {
+    let timeUntilRefreshInMs = 25
+    let response: HTTPResponse
+
+    try {
+      response = await this.requestAccessToken()
+    } catch (err) {
+      // Error without a status code - likely networking, retry
+      this.retryCount++
+      this.lastError = err
+
+      if (this.retryCount % this.maxRetries == 0) {
+        this.tokenEmitter.emit('access_token', { error: this.lastError })
+      }
+
+      this.pollerTimer = setTimeout(
+        this.pollerLoop.bind(this),
+        backoff({
+          attempt: this.retryCount,
+          minTimeout: 25,
+          maxTimeout: 1000,
+        })
+      ).unref()
+      return
+    }
+
+    if (response.headers !== undefined && response.headers.Date != undefined) {
+      const serverTime = Date.parse(response.headers.Date)
+      const skew = Date.now() - serverTime
+      if (this.clockSkewInSeconds == 0) {
+        this.clockSkewInSeconds = skew
+      } else {
+        this.clockSkewInSeconds = (this.clockSkewInSeconds + skew) / 2
+      }
+    }
+
+    // Handle status codes!
+    if (response.status === 200) {
+      let body: any
+      if (typeof response.text != 'function') {
+        this.tokenEmitter.emit('access_token', {
+          error: new Error(
+            'HTTPClient does not implement response.text method'
+          ),
+        })
+        clearTimeout(this.pollerTimer)
+        return
+      }
+      try {
+        body = await response.text() //body?.getReader().read() // TODO: Replace with actual method to get body - needs discussion since different HTTP clients expose this differently (buffers, streams, strings, objects)
+      } catch (err) {
+        // Errors reading the body (not parsing) are likely networking issues, we can retry
+        this.retryCount++
+        this.lastError = err
+        timeUntilRefreshInMs = backoff({
+          attempt: this.retryCount,
+          minTimeout: 25,
+          maxTimeout: 1000,
+        })
+        this.pollerTimer = setTimeout(
+          this.pollerLoop.bind(this),
+          timeUntilRefreshInMs
+        ).unref()
+      }
+      let token: AccessToken
+      try {
+        const parsedBody = JSON.parse(body)
+        // TODO: validate JSON
```

Comment:
This is where an `isAccessToken` typeguard might be useful, if you check the types of the fields as well :smile:

Reply:
So something like 
```
const isAccessToken = (b: object): b is AccessToken => {
  return (
    (b as AccessToken).access_token !== undefined &&
    (b as AccessToken).expires_in !== undefined &&
    typeof (b as AccessToken).access_token === 'string' &&
    typeof (b as AccessToken).expires_in === 'number'
  )
}
```
```
      try {
        const token = JSON.parse(body)

        if (!isAccessToken(token)) {
          throw new Error(
            'Response did not contain a valid access_token and expires_in'
          )
        }

        this.tokenEmitter.emit('access_token', { token })

        // Reset our failure count
        this.retryCount = 0

        // Refresh the token after half the expiry time passes
        timeUntilRefreshInMs = (token.expires_in / 2) * 1000
```
?

Reply:
Yeah, though would set `b` to either `any` or `unknown` (Generally prefer unknown but you can get away with fewer runtime checks using any)
```ts
function isAccessToken(thing: any): thing is AccessToken {
    return Boolean(
        thing && typeof thing === 'object' && 
        typeof thing.access_token === 'string' && 
        typeof thing.expires_in === 'number'
    )
}
```

Reply:
unknown is always better here IMO -- for one thing, the compiler will complain if your type guard is incomplete with unknown.

Reply:
Yeah, need to make a feature request to TypeScript to treat the `typeof x.y === 'string'` as also validating the existence of the field.